### PR TITLE
Logic Definitions are Out of Date

### PR DESCRIPTION
**Pull request by r0bd0g**
Logic Update:

With this first pass I wanted to tackle what I felt were some of the more pressing issues in the logic currently. I have an absolutely gigantic list of new tricks still to parse though. It's gonna be a rough time.

Other Logic PRs that Exist at the Moment of Writing:

- There's another fix currently in PRs for the logic-adjacent issue with King Dodongo's shortcut which I opted not to include here.
- There's also PR to fix a logic deficiency with planted beans in Zora's River, but I felt to implement it as well as I will be touching on other issues with planted beans.
- There's a PR to update the tooltip to a trick in vanilla Jabu with better advice. I agree with the updated advice (for now?) but opted not to make that change here since I'm not really touching on that trick at this time.
---
Assorted Changes:

- Gossip Stone Fairies outside Temple of Time can be spawned with Sun's Song, but only as Adult, because only as an Adult does Sun's Song not reload the area.
- (OHKO Jumpslash Saga Part 1) The Kakariko Watchtower GS jumpslash trick (to get the GS without Bombchus or Slingshot, and subsequent rooftop access as child at night when Man on Roof without Hookshot is also enabled) can be performed without taking fall damage, by letting go from the ladder as low down as possible to still hit the Skulltula. With sticks this just means you have to be more precise with your stick pull and jumpslash timing. The Kokiri Sword has less reach than sticks, so you have to let go at the exact right time mid-climb and while the Skulltula is facing upward so that its hitbox is at its lowest. There are already quite a few tricks that have harder methods to avoid taking damage, and so I felt this was fair to change. I also updated the tooltip to reflect that fall damage is not actually required.
- Added Crater to the Fewer Tunic Requirements tags.
- It's possible the player might want to go from lower to upper Death Mountain Crater, find some area to pass time of day, and then go back from upper to lower to do something time-of-day specific. We missed that use case, so we had no mechanism for logically destroying the boulder while you were still there in lower Crater. You would instead have to have Hover Boots or the double jumpslash trick enabled or something to go back through. I added a check on the exit from upper to lower to allow you back through if you have Hammer and already had access to lower. Crater in general I think has problems, and something ought to be done to fix them, but it's hard to agree on what exactly, and it's causing fixes like this one to keep getting delayed.
- I made a pretty arbitrary change involving leaving the Goron City Grotto. When using the Goron Tunic, if you just run across the lava, you will generally take a single hit before reaching the other side. It's very easy to avoid this by just leaving the ground in any way (backflip, sidehop, jumpslash, whatever) which resets the lava damage timer. We had previously not required knowledge of this fact, and so I felt that it would be more consistent if we didn't start doing it here, even though this is definitely the easiest place where you could use it. There are lots of weird things that could be in logic for crossing this lava, but I think it could get a little out of hand if you tried to be too precise with it, so that's why after looking over it I decided to leave anything else as it was.
- Planted Beans:
-- You can jump from the bean to the Lower ZR PoH or to the Bug Shrub. This is relevant now that child access is not required to plant the bean. I decided to rewrite these lines without checking "is_adult" since if you're not child you must be adult. I put is_adult there originally not really thinking about it, and it was pointed out to me that it wasn't needed, but then some people requested is_adult stay for 'readability', but I'm over it. You're increasing the number of parentheses and that's just asking for trouble. Feel free to disagree, lol. I also implemented it as here(can_plant_bean), which technically works, though only the plant_beans part of that helper is actually relevant. It's kind of overkill but I think it makes more sense written the same way as all of the other bean rides.
-- Bean plant fairies can't be summoned as adult. It used to be that the can_plant_bean helper always checked that you were child, since you had to be child to plant beans, and bean plant fairies relied on this fact to avoid having to check is_child themselves. Under plant_beans this is no longer the case, so I added is_child to all bean plant fairies. I tested if the DMT bean plant fairy could be spawned while the cavern boulder was still there, and it could not be.
-- Riding the DMT Bean normally requires strength or explosives for child to be able to plant it, but this is not longer the case if beans are already planted. I added plant_beans as an alternative to Strength or Explosives, affecting the DMT PoH, the bean trick for lower red rock GS, and DMT Summit access. I considered making DMT Bean planting an event. Certainly there are several ways this could have been implemented.
- Jabu Preboss Ceiling Switch/Skulltula:
-- For the vanilla Jabu ceiling switch, the parantheses were wrong. Luckily it didn't cause an actual bug.
-- Now with boss entrance shuffle, tricks were add to hit the vanilla Jabu preboss switch. One of those tricks allowed using the slingshot or bow from near the entrance into the room. But longshot works just as well and seems to have been forgotten.
-- However, since dungeon shortcuts were added, longshot was in default logic for obtain the skulltula in MQ Jabu through about the same means. I could see it either way, but perhaps having this longshot usage in default logic was a bit of an overreach. Because the two tricks work about the same way, it's been quite a stretch, but I did my best to combine the roughly equivalent tricks for the vanilla preboss switch and MQ skulltula together. The alternative would be to split them into two tricks for MQ and vanilla, only for the reason that it's difficult to succinctly describe them together as the same trick.
-- There's one last thing to mention here though is that the hookshot alone can be used on both the switch and the skulltula, and it's maybe more obvious than using the longshot from over by the door. If you climb up the vines and stand on the right edge of the platform, you can simply aim the hookshot around the glass at your target. The bow and slingshot can both use this exact same method, though it seems a bit harder with those weapons, I think because perhaps Link holds the hookshot a little bit farther to the right.
-- The other trick that was added for the vanilla Jabu preboss switch uses Bombchus. I didn't want to combine the bombchu trick together with the previous trick for bow/slingshot/hookshot because these are quite different methods with different knowledge and skills required. After investigating I found another method of using explosives to hit that switch where adult uses the hover boots to get around the glass, then uses a timed bomb throw to hit the switch. I feel that this method of hitting the switch is novel enough to warrant inclusion as well, and while it too is quite unique from the other methods, at some point I have to start worrying about overdosing on tricks all for the same thing, and so I opted to combine this method together with the bombchu method as part of a "with explosives" trick. (I tried to hit the switch with only bomb bag, no hovers, and it seemed close, particularly as child, but I couldn't quite make it work. It might still be possible, though.)
-- So recap: Fix the brackets. Add longshot to the bow/slingshot trick for the vanilla Jabu switch. Combine MQ Jabu's skulltula into the trick. Reduce the longshot requirement down to just hookshot. Add hovers+bombs to the bombchus trick, renaming it to "with explosives".
- I added comments to MQ Jabu about Lens being unable to be used underwater and adult not being able to swim directly onto a Hookshot pillar due to legs. There was a really confused conversation a while back about what exactly was going on with the logic there.
- (OHKO Jumpslash Saga Part 2) The trick that enables additional weapons to kill the vanilla Forest first room GS can be performed with a jumpslash without taking damage. Generally you use the weird angled let-go technique to kill this skull. If you angle and position it right you can recoil onto the tree branch. So again I removed the OHKO logic and updated the trick's tooltip.
- Momentary Antigrav is no longer banned in restricted glitchless. At the same time, Hookshotting through a huge gap actually is a glitch apparently. These two factors have combined to make the forest temple block room just a gigantic pain. As far as hookshotting through the gap beside the block in MQ forest goes, the trick itself and the logic for it have already been removed, but there are some consequences. I have to consider methods of hitting the switch that weren't relevant when simpler tricks are on and you could just hookshot up there for free.
-- I added a region to vanilla forest for deal with the skulltula in the first hallway, which was long overdue. That doesn't really relate to this I just didn't want it hanging over my head anymore as I was trying to figure the rest of this place out.
-- I updated the jumpslash trick tooltip to refer to the recoil method, and updated the logic to allow any means of jumpslash, not just as adult. There was no need to check for a weapon for the floormaster because a sword item was required to reach this area as child anyway. You have guaranteed access to nuts in order to help calm the babies, but still it is a minimum 5 sticks. That's still manageable enough, and since it's already locked behind a trick I think it's not worth worrying about. And of course the skulltula in this courtyard has been updated to be collected with Boomerang. The comments were a bit too strong about what child couldn't do, so I had to rework those a bit.
-- In MQ forest it could be necessary for adult to defeat the Stalfos for child to pass through. So the logic for beating that Stalfos needed to be wrapped in a here(), except there wasn't actually a region for that so I had to add one for that too.
-- I added a new trick for using the Boomerang to hit the hallway switch. This switch is permanent so it's conceivable that child might have to hit the switch like this for adult to pass through later. Child can also boomerang trick throw the shortcut switch at the top, but since that's logically obsoleted by the new trick, there was no reason to include that.
-- I added the ability for child to perform the recoil to the courtyard ledge in MQ Forest as well. I had something kind of clean before where there was early access to this ledge from both the before after block puzzle regions, but the fact that child can only halfway climb the block puzzle kind of ruined that, so I opted to move everything (except the normal passing through and using small keys route) into the before block puzzle region. This line is absurdly complicated so I will try to explain it to you, and maybe we're just going to pray it's correct because it's really hard to wrap one's mind around.
-- First you need to hit the lower switch to remove the glass blocks. This needs to be an event because one age could potentially open access for the other. I used two events actually, an at() for reaching the switch normally (again, I opted not to include the 'normal' route as part of this logic, so that's why I had to use at() as well instead of one big here()), and a here() event for all methods of jumpslashing the switch through the block.
-- There's the hover boots method of getting in front of the switch, and then the other methods have to hit the shortcut switch at the top in order to jump from the glass blocks. Climbing up normally as adult, or the usual bombchu trick, can be used to hit the switch, but I also included using the Slingshot. It's kind of weird to recognize that you have the option to use the Slingshot there, and I think if it mattered without any tricks it would still be default, but you do have to have tricks enabled before it's relevant, so I don't think it's worth worrying about.
-- Of course to perform the jumpslash from the blocks, you must be able to get on top of them somehow, which, after hitting the switch at the top, requires strength or hookshot, and then you must have a weapon with which to actually jumpslash the hallway switch.
-- Kokiri Sword is too short to reach. Sticks does work and has been included, but they're much more precise than the other methods of using the jumpslash trick. I think the problem is that if you jumpslash even a little bit too early, the stick breaks and becomes too short to hit the switch. It's a bit of a shame that a trick that used to be borderline default logic will now be tied together with this much more precise method, but I think it's the only place it fits and luckily it's not relevant unless the recoil trick is also on.
-- And then finally once the lower switch has been hit, you need some method of getting up there early, either using hookshot, hovers, or the jumpslash recoil trick. Please look this over very carefully and think very hard about whether or not I might have missed anything.
-- In the future I may add a trick for hitting the lower hallway switch with explosives. Only really relevant is throwing bombs at the switch from up top as child. This is an absurdly precise throw and I don't feel completely comfortable adding it as a trick until I understand it better. Bombchuing the switch from below is also relevant as a means to skip sticks, and we don't normally add tricks just to skip sticks or nuts, but since there's room for it by combining it with throwing bombs at the switch, maybe it could be added. I know how to implement this but I just didn't feel it was important enough to include at this time.
- A while back I added some tricks in MQ Fire to skip a requirement to hookshot up from lower to upper 'boulder' maze, as well as needing to raise the hookshot targets to do this. Then to get to the top of the east tower, at first I didn't realize that the switch that spawns those hookshot targets was relevant, but then I noticed that without those targets you couldn't actually jump onto the song of time block. What I failed to notice is that the hookshot target that you're hook up to, to then hook to the elevator, for some reason is ALSO raised by that same switch. So even when using the longshot to skip song of time, the switch to raise the targets still needs to be hit. This was only a problem in the rare case where you would have one of the hookshot skip tricks enabled, but not "Hammer Rusted Switches Through Walls" (which is one of the most commonly enabled tricks). In that case you could have been forced to hammer the rusted switch through the wall despite not enabling that trick. This also means that if for some reason you have that strange trick combination but also the "Pixelshot" trick enabled, pixelshot is relevant here as it completely bypasses that rusted switch. I added pixelshot to the logic, but because this is such an absurd case, I opted to change nothing about the name and tooltip of the pixelshot trick, despite none of that text really making any sense in the context of MQ. This applies both to the "Skull On Fire" and Wall Fairy in this area.
- I changed a hookshot/hovers check in Water Temple to use the Raise Water Level event. I got asked a while back why Hovers/Hookshot was being checked there and so I felt that it would be more clear this way to just use the event that we have for this purpose. The event includes a few more ways of raising the water level back up, but all of those involve middle water level access, which has True access to the cracked wall area already, so this change shouldn't actually change how logic functions in practice.
- I added a trick a while back to skip the small key in the basement of MQ Water, by longshotting to an invisible hookshot target. What I didn't consider at that time is that you could then proceed to reverse through that area of the dungeon to reach the freestanding key, skipping the need for hovers, scarecrow, or the precise jump trick to be enabled. The door back from the Dodongo room is barred, so you still have to defeat the Dodongos. Then after you beat them the door immediately bars back up, as barred doors in rooms you were not intended to reverse through sometimes do, and so you have to reload the room to actually go through. Something that's a bit different compared to other cases of barred doors doing this is that you can't really get a good look at the door closing back up, and so you might assume that door is just always barred if you didn't know any better. So, a bit weird all things considered, but what can you do? I pared back the tooltip a little because I didn't think it needed to be so wordy.
- I changed Morpha and Morpha heart back to using Hookshot rather than Water Temple Clear. Using Water Temple Clear was causing it to appear in the playthrough in every single seed that needed Morpha or Morpha Heart. I think there's some flag somewhere that sets whether or not an Event appears in the playthrough? Not really sure how that works. Was it decided at some point in the past that that event should appear in the playthrough for when you needed to beat Water to swim to Fire Arrows or Fishing Pond?
- I updated the lens trick tooltips for the Shadow Temples to reflect the separation of Bongo without lens from the rest of the lens tricks. Now that the 2nd Dead Hand "can't" spawn out of bounds, I also updated his tooltip to make it more clear that it's just about removing the process of combing the room trying to get him to spawn. I considered making further changes to the logic now that the odds of him falling out of bounds are much lower, but decided not to.
- It was discovered that the trick to light the torches with Din's instead of Fire Arrows in the main room of MQ Spirit can be performed without Bow, and since it's not really any harder than the methods with Bow, I changed the trick to this new method.
- I thought it would make a lot of sense for Ganon's Tower to be part of the Bosses file rather than the Overworld file, so I moved it.
---
Newly-Added Tricks:

- In addition to some of the trick changes and additions above brought about by new settings or ruling changes, I was looking through the trick list for candidates of newly-discovered tricks that I felt were important enough to tackle sooner rather than later (similar to how King Zora Skip was almost immediately added). The first of these is the boulder skip to get from upper to lower in DMC. I pulled some of this directly from Pi's ill-fated DMC rework. I decided it would always require the Goron Tunic except for if you have the Hammer to immediately destroy the boulder, since that boulder is the closest thing but honestly I would expect most people to just do the hammer double jumpslash. I think it'll be a lot cleaner if this new trick cleanly obsoletes the much easier one. I felt that a potential DMC rework was kind of holding up this trick's implementation, and now it's kind of just time to get on with it.
- And the other trick I wanted to add is again in DMC, this time a precise jump to get from lower to central. You will require Hylian or Mirror Shield to shield drop a pot and get an extra push. (I'm going to plug my ears and pretend the fish push method isn't real. I couldn't get it work at all, lol.) This also always requires the Goron Tunic. It would be safe under FTR for reaching Bolero but b/c of the way DMC logic works currently, it would also have to be safe for the bean ride to the PoH, and it just isn't (the closest I could get was about halfway up the bean). DMC rework another time. (Since you need Goron Tunic anyway there was no point to add the same logic again to lower->fire entrance.)
- I added the tricks that I added (or changed the name of, but they were missing) to the hell mode preset. There are still several tricks that aren't in the preset, but I didn't add them because I didn't touch on them when updating the logic.
---
Things I investigated but didn't make changes:

- (OHKO Jumpslash Saga Filler Arc) The Kak Watchtower GS can also be killed with a Boomerang trick throw. I found this throw to actually be pretty annoying to do, though I suspect the average player will find the non-damage jumpslash method harder. Since the Boomerang method is so different from the jumpslash method, I didn't feel that the two methods should be combined into a single trick. And since with the jumpslash trick on the Boomerang method only skips sticks (which are usually not too difficult to find), I didn't feel that it was worth adding a separate trick just for Boomerang.
- It's possible to kill Babas in the basement of MQ Jabu using Deku Shield to reflect the nuts of the Deku Salesman. It's really precise to use it to kill the smaller Baba down below. And it's just too slow and annoying to use to kill the Big Baba to be considered logical.
- The logic has a couple of janky things caused by it not being logical to fight Big Octo as adult. I took another look at the fight, and while it's easy if you know what to do, it's obvious that adult was never intended to go near this boss. The only way to deal damage is with specifically master sword jumpslash. Literally nothing else works.
- (OHKO Jumpslash Saga Part 3) The Spirit Temple Bomb for Light GS can be killed with a jumpslash, and there is a tricky angled let-go technique-or-other to get low enough to avoid fall damage. Because killing this skulltula is not normally part of any trick, we don't usually create tricks specific to OHKO, and the damage-avoidance method is clearly not safe for default, in this case I opted not to implement any logic for a damageless method. It's a bit strange that some of those damage avoidance jumpslashes made it in, but this one doesn't; seems inconsistent. But I also had to juggle be consistent with other things that have been done before, like included more difficult versions of tricks that avoid taking damage (for example -- flame wall skips), and also generally not adding tricks that are specific to OHKO. So I felt that I had to be inconsistent somewhere.

### CODE REVIEW

**Comment by fenhl** [comment]
Fixes #1554.

**Comment by cjohnson57** [comment]
Hellmode preset needs to be updated to include the new tricks and remove the old one that no longer exists

Besides that, as far as I can tell, everything here seems good. Admittedly the MQ forest change is over my head (especially as I don't know MQ) so if someone else could verify I'd feel better about that.

**Comment by r0bd0g** [comment]
I don't normally update things like the changelog or hellmode when I do these logic updates. (I wouldn't be surprised if those things were more out of date than just what's going on here.) I hope that's not a problem? I don't think the rando balks at the least if you try to pass it tricks that don't exist.

Not gonna like the MQ forest one is a bit over my head as well. Definitely the most challenging change in this PR.

**Comment by fenhl** [comment]
The changelog is usually handled by the dev preparing the release, but hell mode should be updated as part of the PR. I can send a PR updating it to your branch after my review if you don't want to update it yourself.

File: SettingsList.py [hunk 1]
```diff
@@ -1502,110 +1538,110 @@ def __init__(self, name, gui_text, min, max, default, step=1,
...
-        'tags'    : ("Lens of Truth","Shadow Temple",),
+        'tags'    : ("Lens of Truth", "Shadow Temple",),
         'tooltip' : '''\
                     Removes the requirements for the Lens of Truth
                     in Shadow Temple MQ for most areas in the dungeon.
                     See "Shadow Temple MQ Invisible Moving Platform
                     without Lens of Truth", "Shadow Temple MQ Invisible
-                    Blades Silver Rupees without Lens of Truth", and
+                    Blades Silver Rupees without Lens of Truth",
                     "Shadow Temple MQ 2nd Dead Hand without Lens of Truth"
```
Comment (fenhl) [inline SettingsList.py:1590]:
```suggestion
                    "Shadow Temple MQ 2nd Dead Hand without Lens of Truth",
```

**Comment by r0bd0g** [comment]
Can you check if anything else is missing from that preset? If there are no new settings and it's only trick changes from this PR, then I will update it.

**Comment by fenhl** [comment]
Not to my knowledge, it's been kept up to date with the latest PRs afaik. If there's something else missing, it can be handled in a separate PR.

**Comment by fenhl** [comment]
> I think there's some flag somewhere that sets whether or not an Event appears in the playthrough? Not really sure how that works.

An event is included in the playthrough by listing it as an item in ItemList.py.

> Was it decided at some point in the past that that event should appear in the playthrough for when you needed to beat Water to swim to Fire Arrows or Fishing Pond?

Probably. As such, it makes sense not to include it for the heart container or medallion locations. But I'm wondering if it should be renamed to `Defeat Morpha` to accomodate boss ER. A similar change is being made in #1531 for Gohma.

**Comment by r0bd0g** [comment]
I spotted the issue with using that event after implementing everything else and genning a few seeds to look for obvious problems. So it was a pretty last-minute change. To be fair I guess, stepping into the blue warp is what matters, not just defeating morpha, maybe 'morpha boss room clear' is the fully accurate way of putting it, so I don't think changing the name of the event matters too much.

File: data/World/Overworld.json [hunk 2]
```diff
@@ -1024,9 +1026,8 @@
             "Kak GS Guards House": "is_child and at_night",
             "Kak GS Tree": "is_child and at_night",
             "Kak GS Watchtower": "
-                is_child and (Slingshot or has_bombchus or 
-                    (logic_kakariko_tower_gs and (Sticks or Kokiri_Sword) and
-                    can_take_damage)) and at_night",
+                is_child and at_night and (Slingshot or has_bombchus or
```
Comment (fenhl) [inline data/World/Overworld.json:1029]:
```suggestion
                is_child and at_night and (Slingshot or has_bombchus or
```

**Comment by r0bd0g** [comment]
There are definitely tricks that I didn't add that aren't in this preset. I'm not sure if they're all from boss entrance shuffle only or if there are more of them.

**Review by fenhl — COMMENTED** [review]
No logic issues spotted, just some typos and trailing whitespace. I didn't review MQ because I'm not familiar enough with it.

**Comment by r0bd0g** [comment]
Don't worry about those conflicts in the OW file. It's just two ways of writing the same fix. You can take whichever way you like better. Now let me try again to break down the trick MQ Forest line. Access to outdoor ledge the usual way by climbing up and falling down is still handled by the old region access. This tricky line is about accessing it early from the block room. To do this you're going to need some way of hitting the hallway switch to remove the glass block. First,
                (at('Forest Temple Outdoor Ledge', True) or
checks if you could go there the normal way. Since it's possible that adult might have to do this to open the way for child to go through and do something. Besides that we need an event to cover all of the ways you might hit the hallway switch.
                    here((logic_forest_mq_hallway_switch_boomerang and can_use(Boomerang)) or
This allows hitting it as child with the Boomerang. Otherwise you will need to use the jumpslash trick.
                        (logic_forest_mq_hallway_switch_jumpslash and
To jumpslash the switch we have to get in front of it somehow. Hover Boots is one way.
                        (can_use(Hover_Boots) or
Besides that we will have to jump from the glass blocks. In order to jump from the glass blocks we're going to need some way of spawning them by hitting the switch at the top of the room. You can use the bombchu trick.
                            (((logic_forest_mq_block_puzzle and has_bombchus) or
Or you can climb to the top as adult and hit the switch directly. Child can climb the whole way up, but from halfway up he can get a line of sight on using the slingshot to hit it. 
                                (Progressive_Strength_Upgrade and (is_adult or Slingshot))) and
Once the switch has been hit, you're going to actually have to get on top of the glass blocks to be able to jump from them, which can be done with either the Hookshot on the targets that you've spawned, or you climb up again using Strength.
                            (Progressive_Strength_Upgrade or can_use(Hookshot)) and
And then once you're on the block you're going to need some method of actually jumpslashing. Kokiri Sword's range is unfortunately too short for this.
                            (is_adult or Sticks)))))) and
Finally with the glass block gone you still need some way to get up to the ledge. You can use Hovers or Hookshot.
                (can_use(Hover_Boots) or can_use(Hookshot) or
Or with Strength to push the blocks, you can then perform the jumpslash recoil trick.
                    (Progressive_Strength_Upgrade and
                    logic_forest_outside_backdoor and can_jumpslash))"
Let me know if anything about this explanation seems unclear.

**Comment by cjohnson57** [comment]
ok, thanks to the breakdown I think I understand it. Thanks r0b!

**Review by cjohnson57 — APPROVED** [review]
